### PR TITLE
Branches C.Gutierrez

### PR DESCRIPTION
**Pull request by CEsGutierrez**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a spotlight method that finds the highest-voted work in the collection.
Describe how you approached testing that model method. What edge cases did you come up with? | I mostly tested it in the terminal and the console. I could not figure out syntax to identify the specific instance variable that was produced as the solution.  
What are `session` and `flash`? What is the difference between them? | Session is the persistence of user's data. Flash allows for messages to display during 1 or 2 request/response cycles. 
What was one thing that you gained more clarity on through this assignment? | I have a far greater appreciation for testing and formatting. I feel that a lot of this project was about implementing new strategies which led to unexpected results for both display and functionality. 
What is the Heroku URL of your deployed application? |  https://m-ranker.herokuapp.com/ 
Do you have any recommendations on how we could improve this project for the next cohort? | I had trouble following the Wave descriptions for this project, which made it challenging to measure my progress.

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | no
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | no
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | no
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | only when logging in a second time (doesn't when the user is created)
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes, but doesn't log user in
Logging in as an already existing user has a specific flash message | no

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | your routes would be more controlled if you chose a different way to do routes for users and votes that wasn't `resources` with no limitations
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | yes, although there's opportunity for more
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes, in WorksController
Some controller, likely the `WorksController`, has a controller filter for finding a work | no
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes, although I would comment that I think the `create` action has room for growth in clarity

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | just validations
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | just validations
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | just validations
Work model has tests with a section on all business logic methods in the model, including their edge cases | work in progress

## Overall Feedback


Good work on this project! I'm really happy that complex features like logging in and voting work smoothly on your project.


My biggest concern for your project is: Creating a new work and editing a work doesn't work on your project!
1. Why doesn't creating work correctly? From glancing at the code, it seems like your form is trying to say that a category is either `1`, `2`, or `3`, when your work validations say that's invalid. Beyond that, your WorksController gives back the message `"Two works in the same category can't have the same name."`
2. Why doesn't editing work correctly? From glancing at the code, I think it's related to the category thing too


My next biggest concerns are to make sure you got as much practice with model testing as possible, and your model tests were in progress, so let me know what questions you have there, since we won't get much practice anymore!


In terms of code style, I'd say that your MVC code could use love.
- I think that the `find_spotlight` and the `find_top`... methods in the model could use a healthy refactoring to simplify that you may see if you give a fresh look at that code
- I think that some of the logic in the view (particularly around making a table based off of the category) could be moved to a view helper, controller, or model! For example, in the controller, there could be code like `@albums = Work.where(category: "album")`, so you don't have to chec, in the view things like `<% if work.category == "album"%>` in the works/index view
- I think in your controllers, particularly the `create` action, you added some pretty specific logic about validations, which could be made more generic


If you had more time to refactor, I would be looking for refactorings like partials, controller filters, etc




That being said, I also I want to apologize that this feedback came so late. It's been two weeks and I know your understanding of things has likely changed dramatically. That being said, we'll check in with each other about this, just to make sure that we cover all of the questions we can!!

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    username = params[:user][:username]
+    @user = User.find_by(username: username)
+    
+    if @user 
+      session[:user_id] = @user.id
+      flash[:message] = "Successfully logged in as #{@user.username}"
+      redirect_to root_path
+    else
+      @user = User.new(username: params[:user][:username])
+      session[:user_id] = @user.id
```
Comment (tildeee) [inline app/controllers/users_controller.rb:44]: Logging in as a new user doesn't work on the first log in because you set `session[:user_id] = @user.id`, but `@user` hasn't been saved in the DB so it doesn't have an ID
